### ISSUE TITLE
[Linux/Proton] Overloader: IMGs cannot be symbolic links

### ISSUE THREAD

**Issue by purplebar0** [ThirteenAG/GTAIV.EFLC.FusionFix#324]
IMGs have to be placed into `update` as actual files, using links fails to load them. It may be a Wine/Proton quirk, but I suspect it could also be an issue on Windows.

Given that I like to organise my mods, having to stuff everything inside one tree is a bit of a turnoff, but luckily it was simple enough to link everything together [like this](https://gtaforums.com/topic/934545-fusionfix/?do=findComment&comment=1072316540). However, trying to do the same for IMGs didn't work, the modified files within didn't load.

---

Using GTAIV 1.0.7.0 on Proton Experimental

**Comment by Tomasak** [comment]
img loading unlike rest of files loaded from update folder use something else use some extra hooking.
however mod is designed mainly for latest CE version and what linux does when running windows apps is also out of scope for this issue list.

**Comment by ThirteenAG** [comment]
I think I removed a file check from img loader, and it needs to be removed from UAL as well. I should include it in the next release.

**Comment by ThirteenAG** [comment]
I've removed this restriction in latest UAL version.

**Comment by ThirteenAG** [comment]
I appears at bottom left corner of the menu.

**Comment by purplebar0** [comment]
> I appears at bottom left corner of the menu.

It probably counts as a separate issue, but even if IMGs load as regular files, I'm not seeing it...

![2024-01-16-135757_1920x1080_scrot](https://github.com/ThirteenAG/GTAIV.EFLC.FusionFix/assets/107264962/af491c84-1841-4636-8720-2de35fbd5f22)

In any case, I can instantly tell if the IMG has loaded since it modifies one of the phone themes, so for the purpose of this issue, the "extra info" isn't too important.

**Comment by ThirteenAG** [comment]
You need to wait until PAUSED text appears. Have you tried to replace original img file with symlink? Does that work?

**Comment by ThirteenAG** [comment]
>  nothing appears to change.

Well, it will only show in complete edition.

>  works as a regular file, doesn't work as a symlink.

I get that in update folder, but what about files in original folder? If the game can't load them either, it's not fusion fix's loader issue.

**Comment by purplebar0** [comment]
Oh! Right. My file structure used to look like this:

```
GTAIV
	\ update_mods
		\ (mods go here)
	\ update
		\ (dirs and symlinks to paths under update_mods go here)
```

Now I changed it to this and everything works as expected:

```
GTAIV
	\ update
		\ __orig
			\ (mods go here)
		\ (dirs and symlinks to paths under update_mods go here)
```

**Comment by ThirteenAG** [comment]
Forget about the update folder, replace e.g. `"GTAIV\TBoGT\pc\models\cdimages\vehicles.img"` with a symlink, will it work like that?
Because currently I don't see anything preventing symlinks to be loaded, unless the game doesn't load them (for w/e reason).

**Comment by purplebar0** [comment]
> Forget about the update folder, replace e.g. `"GTAIV\TBoGT\pc\models\cdimages\vehicles.img"` with a symlink, will it work like that? Because currently I don't see anything preventing symlinks to be loaded, unless the game doesn't load them (for w/e reason).

Yep, works whether I link it in the same directory, or some arbitrary place the game wouldn't normally look, like `GTAIV\topsecret`. I did make sure that FF didn't overwrite it by renaming the `update` folder temporarily.

**Comment by purplebar0** [comment]
Well, this is bizarre: the examples from [my older comment](https://github.com/ThirteenAG/GTAIV.EFLC.FusionFix/issues/324#issuecomment-1893729283) appear to work completely differently.

If I organise files the former way (linking to a separate directory), everything is fine. However, doing it the latter way (having a subdirectory store all the files), it completely ruins the behaviour for limiting files to epsiodes.

Loading a new game of TLAD with FusionFix applied, for example, loads the first IV mission, but the player model is Johnny and all the UI is grimey like TLAD.

**Comment by ThirteenAG** [comment]
Strange, what about non img files then, does ual loads symlinks from update folder?

**Comment by purplebar0** [comment]
> Strange, what about non img files then, does ual loads symlinks from update folder?

Nope. Using the aforementioned "former approach", I noticed that modified `radio_hud_noncolored.wtd` (a loose file) hasn't loaded.

I suspect that FF's behaviour of determining what is to be loaded by which episode might be the issue at hand.

Maybe it'd be best to just switch to using hard links, I totally understand that symlinks become tricky to deal with in such situations.

**Comment by ThirteenAG** [comment]
I will check filesystem api behavior for this later.

**Comment by ThirteenAG** [comment]
[GTAIV.EFLC.FusionFix.zip](https://github.com/ThirteenAG/GTAIV.EFLC.FusionFix/files/13960666/GTAIV.EFLC.FusionFix.zip)

I found an issue preventing symlinks from loading, try this build for imgs. As for regular files, I can see that UAL loads symlinks fine. I will double check.

**Comment by purplebar0** [comment]
> https://ci.appveyor.com/api/buildjobs/c63lo84y4uccypts/artifacts/GTAIV.EFLC.FusionFix.zip
> 
> I found an issue preventing symlinks from loading, try this build for imgs. As for regular files, I can see that UAL loads symlinks fine. I double checked, it's working.

I tested the build and indeed, links pointing outside of `update` seem to work, but the per-episode loading is still broken. Should I open another issue for that?

**Comment by ThirteenAG** [comment]
[GTAIV.EFLC.FusionFix.zip](https://github.com/ThirteenAG/GTAIV.EFLC.FusionFix/files/13972095/GTAIV.EFLC.FusionFix.zip)

Try this.

**Comment by purplebar0** [comment]
> [GTAIV.EFLC.FusionFix.zip](https://github.com/ThirteenAG/GTAIV.EFLC.FusionFix/files/13972498/GTAIV.EFLC.FusionFix.zip)
> 
> Try this.

Tested all 3 scenarios:

- IMGs as regular files
- IMGs as links outside of `update`
- IMGs as links inside of `update`

In all of these cases, IMGs appear to load, but starting a new TLAD game starts The Cousins Bellic as observed before.

**Comment by ThirteenAG** [comment]
> but starting a new TLAD game starts The Cousins Bellic as observed before.

What does that mean? Img loader shouldn't affect that.

**Comment by purplebar0** [comment]
> > but starting a new TLAD game starts The Cousins Bellic as observed before.
> 
> What does that mean? Img loader shouldn't affect that.

Sorry, I realised I referred to this effect in different ways in previous comments. "IV in TLAD"/playing first mission as Johnny meant the same thing as I described in that comment. As in, IMGs aren't properly isolated to episodes.

**Comment by ThirteenAG** [comment]
I don't observe this behavior in debugger. Are you sure you didn't accidentally place the img into wrong folder?
Or how can reproduce this?

**Comment by ThirteenAG** [comment]
I still don't understand, I've made a symlink at `GTAIV\update\GTAIV.EFLC.FusionFix\TBOGT\TBoGT.FusionFix.img`, the actual file is on desktop, and it loads only when tbogt is loading.

**Comment by purplebar0** [comment]
> I still don't understand, I've made a symlink at `GTAIV\update\GTAIV.EFLC.FusionFix\TBOGT\TBoGT.FusionFix.img`, the actual file is on desktop, and it loads only when tbogt is loading.

If it helps figure out the issue:
- TLAD and TBoGT are in root directory of the game (ZolikaPatch `LoadDLCs=1`, sadly there appears to be no alternative)
- XLiveLessAddon is installed, only `RemoveRegistryPathDependency` and `SkipWebConnect` are enabled
- ColAccel is installed, but I am confident it has no effect on this

If nothing appears unusual, I am going to attempt to get CE running and retest.

**Comment by ThirteenAG** [comment]
> Directories/subdirectories are not links.

Actually I just fixed as well. Code wise everything looks good to me now, I don't think these issues you have now are related to img loader.

**Comment by purplebar0** [comment]
Unfortunately, since the changes were pushed to the latest release, I need to rely on 2.4.9 to use IMGs (even as regular files, [scenario 1](https://github.com/ThirteenAG/GTAIV.EFLC.FusionFix/issues/324#issuecomment-1898175201)) at all.

This might now be a general issue for configurations of old versions (pre-CE) where DLCs are in the root directory and ZolikaPatch `LoadDLCs=1` is used.

**Comment by ThirteenAG** [comment]
Xliveless Addon has all the code for DLC support, everything works correctly there.
